data-prepper-script/src/main/java/org/opensearch/dataprepper/script/parser/LogListener.java

```diff
@@ -0,0 +1,230 @@
+/*
+ * Copyright OpenSearch Contributors
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+package org.opensearch.dataprepper.script.parser;
+
+import org.antlr.v4.runtime.ParserRuleContext;
+import org.antlr.v4.runtime.Token;
+import org.antlr.v4.runtime.tree.ErrorNode;
+import org.antlr.v4.runtime.tree.TerminalNode;
+import org.opensearch.dataprepper.script.antlr.DataPrepperScriptListener;
+import org.opensearch.dataprepper.script.antlr.DataPrepperScriptParser;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.util.Collections;
+import java.util.List;
+import java.util.stream.Collectors;
+import java.util.stream.Stream;
+
+/**
+ * @since 1.3
+ * LogListener is a utility listener that logs every event enter and exit. Useful for debugging and developing new
+ * listeners.
+ */
+public class LogListener implements DataPrepperScriptListener {
+    private static final Logger LOG = LoggerFactory.getLogger(LogListener.class);
+
+    private int level = 0;
+
+    /**
+     * @since 1.3
+     * Utility function to repeat String str, n number of times.
+     * @param n is the number of times str will be printed
+     * @param str the string to be repeated.
+     * @return String containing n instances of str
+     */
+    private String nCopiesOf(final int n, final String str) {
+        return String.join("", Collections.nCopies(Math.max(n, 1), str));
+    }
+
+    /**
+     * @since 1.3
+     * Get self or child nodes of type Terminal Node (Indivisible Rule/Token)
+     * @param ctx Current parsing context
+     * @return List of self or all child nodes that are Terminal Nodes. List may be empty if no terminal nodes found.
+     */
+    private List<TerminalNode> getTerminalNodes(final ParserRuleContext ctx) {
+        if (ctx instanceof TerminalNode) {
+            return Collections.singletonList((TerminalNode) ctx);
+        }
+        else if (ctx.children == null) {
```

Comment:
If there is a syntax error in the statement being parsed a node that should have children may not. This is used to avoid throwing a null pointed exception in those cases.